frontend/src/app/main/data/workspace/libraries.cljs

```diff
@@ -698,3 +707,82 @@
         new-pos            (gpt/add root-pos delta)]
     new-pos))
 
+(defn- generate-sync-colors-file
+  [state file-id]
+  (let [colors
+        (if (nil? file-id)
+          (get-in state [:workspace-data :colors])
+          (get-in state [:workspace-libraries file-id :data :colors]))]
+    (when (some? colors)
+      (loop [pages (seq (vals (get-in state [:workspace-data :pages-index])))
+             rchanges []
+             uchanges []]
+        (let [page (first pages)]
+          (if (nil? page)
+            [rchanges uchanges]
+            (let [[page-rchanges page-uchanges]
+                  (generate-sync-colors-page file-id page colors)]
+              (recur (next pages)
+                     (concat rchanges page-rchanges)
+                     (concat uchanges page-uchanges)))))))))
+
+(defn- generate-sync-colors-page
+  [file-id page colors]
+  (let [linked-color? (fn [shape]
+                        (some
+                          #(let [attr (name %)
+                                 attr-ref-id (keyword (str attr "-ref-id"))
+                                 attr-ref-file (keyword (str attr "-ref-file"))]
+                             (and (get shape attr-ref-id)
+                                  (= file-id (get shape attr-ref-file))))
+                          cp/color-sync-attrs))
+
+        linked-shapes (cph/select-objects linked-color? page)]
+    (loop [shapes (seq linked-shapes)
+           rchanges []
+           uchanges []]
+      (let [shape (first shapes)]
+        (if (nil? shape)
+          [rchanges uchanges]
+          (let [[shape-rchanges shape-uchanges]
+                (generate-sync-colors-shape shape page colors)]
+            (recur (next shapes)
+                   (concat rchanges shape-rchanges)
```

Comment:
Don't use concat in this case, is much better using `into` or `app.common.data/concat`. The builtin clojure concat work with seqs (not with vectors) so it is lazy an incurr additional overhead (is worth in other cases but in loop and similar procedimens you just want `into` or `d/concat`.